ott/core/initializers.py

```diff
@@ -0,0 +1,330 @@
+  regularized potential
+
+  Args:
+    vector_min: Use vectorized inner loop if true. Defaults to True.
+    tol: DualSort convergence threshold. Defaults to 1e-2.
+    max_iter: Max DualSort steps. Defaults to 100.
+    stop_gradient: Do not trace gradient. Defaults to True.
+  """
+
+  def __init__(
+      self,
+      vector_min: bool = True,
+      tol: float = 1e-2,
+      max_iter: int = 100,
+      stop_gradient: bool = True
+  ) -> None:
+
+    super().__init__()
+
+    self.tolerance = tol
+    self.stop_gradient = stop_gradient
+    self.max_iter = max_iter
+    self.update_fn = lambda f, mod_cost: jax.lax.cond(
+        vector_min, self.vectorized_update, self.coordinate_update, f, mod_cost
+    )
+
+  def vectorized_update(
+      self, f: jnp.ndarray, modified_cost: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Inner loop DualSort Update.
+
+    Args:
+      f : potential f, array of size n.
+      modified_cost: cost matrix minus diagonal column-wise.
+
+    Returns:
+      updated potential vector, f.
+    """
+    f = jnp.min(modified_cost + f[None, :], axis=1)
+    return f
+
+  def coordinate_update(
+      self, f: jnp.ndarray, modified_cost: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Coordinate-wise updates within inner loop.
+
+    Args:
+      f: potential f, array of size n.
+      modified_cost: cost matrix minus diagonal column-wise.
+
+    Returns:
+      updated potential vector, f.
+    """
+
+    def body_fn(i, f):
+      new_f = jnp.min(modified_cost[i, :] + f)
+      f = f.at[i].set(new_f)
+      return f
+
+    return jax.lax.fori_loop(0, len(f), body_fn, f)
+
+  def init_sorting_dual(
+      self, modified_cost: jnp.ndarray, f_potential: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Run DualSort algorithm.
+
+    Args:
+      modified_cost:  cost matrix minus diagonal column-wise.
+      f_potential: potential f, array of size n.
+
+    Returns:
+      potential f, array of size n.
+    """
+
+    def body_fn(state):
+      prev_f, _, it = state
+      f_potential = self.update_fn(prev_f, modified_cost)
+      diff = jnp.sum((f_potential - prev_f) ** 2)
+      it += 1
+      return f_potential, diff, it
+
+    def cond_fn(state):
+      _, diff, it = state
+      return jnp.logical_and(diff > self.tolerance, it < self.max_iter)
+
+    it = 0
+    diff = self.tolerance + 1.0
+    state = (f_potential, diff, it)
+
+    f_potential, _, it = jax.lax.while_loop(
+        cond_fun=cond_fn, body_fun=body_fn, init_val=state
+    )
+
+    return f_potential
+
+  def init_dual_a(
+      self,
+      ot_problem: linear_problems.LinearProblem,
+      lse_mode: bool,
+      init_f: Optional[jnp.ndarray] = None,
+  ) -> jnp.ndarray:
+    """Apply DualSort algo.
+
+    Args:
+      ot_problem: OT problem.
+      lse_mode: Return potential if true, scaling if false.
+      init_f: potential f, array of size n.
+
+    Returns:
+      potential f, array of size n.
+    """
+    cost_matrix = ot_problem.geom.cost_matrix
```

Comment:
This will fail if `ot_problem.geom.cost_matrix` is an online point cloud, as it won't be materialized. Would produce a meaningful error message here.

Reply:
I'd say throwing an error is better here, since it will be thrown by some invalid operation (like `None + <some_number>`) later anyway.

Reply:
Currently if is online it will return the default potential, so will work just not be the initializer specified. 

I guess the option is to run regardless  with a warning or return an error and stop

Reply:
I'd be fine with both options, though slightly prefer raising an error, since user needs to specify this initializer explicitly and the warning can be easily ignored (+ is erased after jitting). 
Pinging @marcocuturi for opinion.

Reply:
Spoke with Marco today, error is preferred, though in the future we might consider handling online geometries.

Reply:
Sure, will add this